cmd/gardener-extension-shoot-dns-service/app/app.go

```diff
@@ -106,9 +110,64 @@ func (o *Options) run(ctx context.Context) error {
 		return fmt.Errorf("could not add controllers to manager: %s", err)
 	}
 
+	if err := (&migrations{}).AddToManager(ctx, mgr); err != nil {
+		return fmt.Errorf("could not add migrations to manager: %s", err)
+	}
+
 	if err := mgr.Start(ctx); err != nil {
 		return fmt.Errorf("error running manager: %s", err)
 	}
 
 	return nil
 }
+
+type migrations struct {
+	client client.Client
+	log    logr.Logger
+}
+
+var _ manager.Runnable = &migrations{}
+var _ manager.LeaderElectionRunnable = &migrations{}
+
+func (m *migrations) NeedLeaderElection() bool {
+	return true
+}
+
+func (m *migrations) Start(ctx context.Context) error {
+	// TODO (Martin Weindel) can be deleted after release 0.55.0
+	return m.deleteObsoleteManagedResources(ctx)
+}
+
+func (m *migrations) AddToManager(_ context.Context, mgr manager.Manager) error {
+	m.client = mgr.GetClient()
+	m.log = mgr.GetLogger().WithName("migrations")
+	return mgr.Add(m)
+}
+
+func (m *migrations) deleteObsoleteManagedResources(ctx context.Context) error {
+	log := m.log.WithName("deleteObsoleteManagedResources")
+
+	list := &metav1.PartialObjectMetadataList{}
+	list.SetGroupVersionKind(resourcesv1alpha1.SchemeGroupVersion.WithKind("ManagedResourceList"))
+	if err := m.client.List(ctx, list); err != nil {
```

Comment:
Does it make sense to use `.metadata.name` field selector here?

Reply:
Using `.metadata.name` field selector requires to add a field indexer for the cache with lengthy additional code. I don't think it is worth the effort as it runs only once on startup.